### DISCUSSION TITLE
Document key-priorieties for development

### DISCUSSION THREAD

**Discussion by 6543** [woodpecker-ci/woodpecker#2152] [General]
1. provide sensible defaults
2. provide good and consistent documentation
3. make sure to write generic func and split up big func into sub functions
4. use packages for each topic and if to big split asap in subtopics

---
https://github.com/woodpecker-ci/woodpecker/issues/1860#issuecomment-1590175406

**Comment by anbraten** [comment]
I disagree with the second point at least with its priority. In case of configs its more important to have explaining configs instead of "short" configs doing "magic" as for example DRY is a programming concept, but repeating in configs is fine and most of the time way more readable IMO. I would say configs should be intuitive and self explaining instead.

**Reply by 6543** [reply]
Updated

**Comment by 6543** [comment]
Good point

**Comment by qwerty287** [comment]
@6543 did most of this with https://github.com/woodpecker-ci/woodpecker/pull/3184, but I don't think we can close this.

Some points which need to be discussed (I think):
1. Which features do we implement/accept? Currently, we accept almost all features, even though if they are really rare usecases.
2. Support for backends: Currently, we state WP on our website kind of as "container-first" which is wrong. However, there are a lot of features just usable with `docker` (see https://github.com/woodpecker-ci/woodpecker/discussions/3191, https://github.com/woodpecker-ci/woodpecker/pull/3156).
3. What should be supported out-of-the-box, what should be moved to addons? Esp. with these in mind:
    - https://github.com/woodpecker-ci/woodpecker/pull/3179
    - Which new backends are allowed to be added to the core repo? (Just take a look at https://github.com/woodpecker-ci/woodpecker/discussions?discussions_q=is%3Aopen+label%3Abackend)
    - Which new forges are allowed to be added to the core repo?

**Reply by zc-devs** [reply]
I'm a bit concerned about this `Which features do we accept` and `Which new forges are allowed to be added to the core repo`.
Why this should not be allowed if contributor make... a contrubution? Do project have a lot of contributors / a deep queue of contributions to choose from them?
And I have feeling, that whole addons story (will) tears apart the community. Shouldn't we consolidate efforts?

**Reply by qwerty287** [reply]
> Why this should not be allowed if contributor make... a contrubution? Do project have a lot of contributors / a deep queue of contributions to choose from them?

My main issue with this is that making a contribution is not everything that must be done. That's just the first part. Adding a new feature means this feature must be maintained and in many cases, this does not happen from the original contributors, so we as maintainers must do this, even though we don't use the feature.

Splitting this and building the software modularly is a way to prevent issues like this.

Putting everything into the core means that you download a binary/image that contains hundreds of features you don't use. Besides the issue that the compile time and binary size increases, this can also increase the impact of security issues etc.

If there's a simple way about adding a feature using some kind of addon or extension approach, what are important downsides here?

Of course, if a feature is deeply integrated into the core and can't be implemented using an addon approach, I don't have an issue with adding it to the core. Similar if the feature is a really simple addition.

> And I have feeling, that whole addons story (will) tears apart the community. Shouldn't we consolidate efforts?

Even though they were implemented differently originally, we currently only plan to support addons for forges.
Not sure if you knew this, but I don't really get your point here. Can you elaborate a bit more?

**Reply by zc-devs** [reply]
> we as maintainers must do this, even though we don't use the feature

Why `must`? I, personally, do not expect it. If you do not use Kubernetes, I do not expect support from you. If I do not use docker/local it is strange to expect implementation for that backend from me. Even if I agree to make some changes, that will be untested (from my side). Sounds logical to me.
Consider current situation. What backends do maintainers use (owners, because only they are active maintainers)? Docker, I suppose. Who uses Kubernetes? Who maintains Local, except 6543? So, Docker in core as well as Local luckily. Is Kubernetes going to be an addon? (No, saw the last answer. But the main point is about responsibility and expectations).

> contains hundreds of features you don't use

What's the problem? I do not use hundreds of features in Gitea, for example.

> binary size increases

Binary size? 15 MB? Are you kidding?
v1.0.0 - 13.03 MB
v2.0.3 - 15.17 MB
Has it become an issue already?

> compile time

From my experience: compile time is nothing compared to downloading build images (xgo, particularly) and dependencies. Compilation takes 3 minutes on one core of old Intel Celeron. And multiple cores don't do it much more faster, BTW.

> security issues

Get it, agree.

> If there's a simple way about adding a feature using some kind of addon or extension approach, what are important downsides here?

If there is, maybe. One does not simply ~walk into Mordor~ make an addon. To me it's more simpler to make changes on top of main codebase. However, I'm curious how addons will go.

> we currently only plan to support addons for forges. Not sure if you knew this

No, didn't. 

> Shouldn't we consolidate efforts? Can you elaborate a bit more?

Considering previous point, I thought...
1. If all functionalities is going to be addons, then what is going to be the core?
2. If we have to go addons/fork approach, then it `tears apart the community`.

--- 

Edit 1

> Adding a new feature means this feature must be maintained and in many cases, this does not happen from the original contributors

Someone is willing to implement some feature. For what? Just for fun? Seems, it desired and he uses it (will).
Then one are going through ~hell~ merging process. Finally, PR got merged.
Then what does contributor do? Stops using feature? Not likely.
Expects, that following subrequest will be done by maintainers? Then it should be told clearly, that it ~is not gonna~ *may* happen. But, I think, it follows from License.

**Reply by qwerty287** [reply]
> Why `must`? I, personally, do not expect it. If you do not use Kubernetes, I do not expect support from you.

That's what you should assume as user, but there are a lot of users thinking differently.

> owners, because only they are active maintainers

Nah, that's not true. Almost all maintainers are active, maybe not as active as the owners, but they do reviews and contributions (some also for other repos the the core, ex. pat-s on the helm chart and a lot of plugins).

> Is Kubernetes going to be an addon? (No, saw the last answer. But the main point is about responsibility and expectations).

Kubernetes has enough users and external contributions and also some maintainers do use it. Maybe it's not used as widely as docker/local, but it's used.
As an example of the opposite site, see the ssh backend which was removed in v2 because it was completely broken, but nobody - neither maintainers nor external contributors - fixed something there.

> What's the problem? I do not use hundreds of features in Gitea, for example.

Me neither, and I don't like that gitea is that bloated.

> Binary size? 15 MB? Are you kidding?
> v1.0.0 - 13.03 MB
> v2.0.3 - 15.17 MB
> Has it become an issue already?

No, but I think it's better to keep it as small as possible.
Of course it's not big yet, but for example gitea has iirc about 120 mb which is too much in my opinion.

The point is actually more about code size, I wasn't clear here, sorry.
Adding more and more features means adding more and more code which makes it harder to get an overview. Even though I contribute to woodpecker for almost two years now, there is still enough code I didn't fully understand yet. Probably, this can be changed with better docs and a more clean code structure, but we have a lot of old stuff from drone that must be refactored.

> > Shouldn't we consolidate efforts? Can you elaborate a bit more?
> 
> Considering previous point, I thought...
> 1. If all functionalities is going to be addons, then what is going to be the core?

Nobody says "all functionality will be addons". The addon feature should rather be a way to allow users to implement features on their own because we reject the proposal as almost nobody needs it (or for whatever reason).

> Just for fun?

Well, I do this sometimes :)
I get your point here though, but - maybe this does not apply to new features but rather to docs changes, so feel free to ignore this - there are also events like hacktoberfest or just people wanting to contribute to some open source project, so they might not need a feature.

> Then what does contributor do? Stops using feature? Not likely.

Maybe, this does not happen directly after the feature was added, but if your requirements change, this is nothing unusual I think.
If a feature then has no users anymore because it's too specific, it won't get any maintenance anymore. Such features should be implemented with addons instead. If the user don't need it anymore, just don't use the addon anymore.
Of course, features that are used by many users wouldn't have this issue.

> Expects, that following subrequest will be done by maintainers?

What do you mean with "subrequest"?

**Reply by 6543** [reply]
So my two cents about feature contributions:

If it is a unicorn one shot thing: let it be an addon
If it would violate our core dev gidelines: also addon

Else: just let it be merged into our codebase and document it as experimental.
If it sees high usage or contributions, we should considder it usefull and mark it as stable if we are confident enouth that it will stay.

Best example is the ssh backend ... witch we accepted but removed after no maintinance (never leafed experimental ...)

**Reply by 6543** [reply]
TLDR: we should encourage contributions into our project ... and acept such too

**Reply by qwerty287** [reply]
That's basically what I'd say too.

> If it would violate our core dev gidelines

Now the question would be what "core dev guidelines" are. Because if I think about forges, I'd say a forge must at least support these features:
- OAuth2
- Hooks

to be merged into our core.

But let's not discuss this here, but rather use a new discussion if we finished discussion about the general points.

**Reply by 6543** [reply]
Agree :)

PS: i created https://github.com/woodpecker-ci/woodpecker/issues/3364

Feel free to summarize things a bit as code comment, else just wait and review the contribution addressing this issue :)

**Reply by zc-devs** [reply]
> Almost all maintainers are active...

That is good. It was just my feeling about main repo, because addons was/is about main repo.

> Kubernetes has enough users and external contributions and also some maintainers do use it

Now. I chose it even though it was experimental. I invested my efforts (as others) and look at it now.: sometimes it beats even Docker :) What if it was some third-party addon?

> ssh backend which was removed in v2

Was it hard? It looks like click-click-delete to me, just removing a couple of packages. More managerial work than technical.

> I don't like that gitea is that bloated

I was concerned too.
- But first of all, nobody stays behind me with a gun and push me to use features.
- Secondly, it is designed for small-meduim teams, I think.
I started using image registry in Gitea. It was as simple as click-enable, push, pull. Awesome. Then I moved to Harbor.
If I want to use software packages registry, I'll probably start with Gitea because of simplicity. And then, probably, move to Nexus, for example.
- As an admin I dont have to search addons, plugins and so on. I just go to admin console, enable LDAP/OAuth, Issues, Wiki. Just click-click and done. I dont have to deploy Jira, Confluence or whatever.

> gitea has iirc about 120 mb which is too much

[Gitea 1.21.5 - 65.21 MB](https://hub.docker.com/layers/gitea/gitea/1.21.5/images/sha256-ad0c0ed72ddf7cc56b45b32178a0c0e8b6527ca0083f9e5288d93829c5702508?context=explore)

I have nice example: [nextcloud - 404.44 MB](https://hub.docker.com/layers/library/nextcloud/stable-apache/images/sha256-d401d74a620aebd9572a1adb5a983698d2bc1ae30816db7a176a5751890253d1?context=explore). They have modules, cool. But all core modules in the image. Yet, [users have to build own images](https://github.com/nextcloud/docker/issues/1638). You know what after all? I dont care about half GB size. I'm not on dialup or GPRS and can afford a couple of minutes to download image. And it happens once per month (?). I just use it.

> clean code structure

This. Agree.

> I do this sometimes
> they might not need a feature

Me too. But it is not just random thought from my mind. I need it / would like to use it. Or someone need it and I just like it and curious to implement. There is always request (issue).

> requirements change, this is nothing unusual

Of course. Life flows. It always changes. See comment about Gitea.
But should it be a reason to reject contribution?
Let's consider Bitbucket Server / Kubernetes. I'm evaluating the project.
When was it started? Pretty long time ago (+).
Is it active? Yes (+).
Is it one/two persons project? No (+), Yes(-).
Does it support Kubernetes without DinD? Yes (+), but experimental. OK, will see how it goes. Oh, there is couple of bugs... I can fix / workaround that.
Does it support Bitbucket Server? No (-). Not even buggy implementation that I can fix.

> What do you mean with "subrequest"?

Like I were merged libvirt/podman/etc initial/experimental support and `It's not my problem anymore. I want this and this. Let maintainers do it now`.
I'm alone here and have to rely on myself. There is a chance that I won't be alone. But if that initial implementation wasn't merged, chances are going down.

> we should encourage contributions into our project ... and acept such too

You are way more better explained my thoughts than I 🤣 

---

Sorry for long post and offtopic sometimes.

**Reply by qwerty287** [reply]
> Now. I chose it even though it was experimental. I invested my efforts (as others) and look at it now.: sometimes it beats even Docker :) What if it was some third-party addon?

Sure, we need a starting point for some features. But take a look at the original issue: https://github.com/woodpecker-ci/woodpecker/issues/9. It has 12 :heart: reactions so it was kinda clear that this feature will have a lot of users.

> [Gitea 1.21.5 - 65.21 MB](https://hub.docker.com/layers/gitea/gitea/1.21.5/images/sha256-ad0c0ed72ddf7cc56b45b32178a0c0e8b6527ca0083f9e5288d93829c5702508?context=explore)

I mainly mean the binary directly:

![grafik](https://github.com/woodpecker-ci/woodpecker/assets/80460567/786f4ab3-36c0-4890-bf9e-727a72d0d9f9)

---

However, it looks like in general we all agree to the same points @6543 wrote above:
> If it is a unicorn one shot thing: let it be an addon
> If it would violate our core dev gidelines: also addon
> 
> Else: just let it be merged into our codebase and document it as experimental.
> If it sees high usage or contributions, we should considder it usefull and mark it as stable if we are confident enouth that it will stay.

(core dev guidelines do not exist yet, so this is next discussion point)
